lib/racecar/runner.rb

```diff
@@ -214,6 +219,26 @@ def process_batch(messages)
       end
     end
 
+    # librdkafka will continue to try to deliver already queued messages, even if ruby-rdkafka
+    # raised before that. This method detects any unrecoverable errors and resets the producer
+    # as a last ditch effort.
+    # The function returns true if there were unrecoverable errors, or false otherwise.
+    def reset_producer_on_unrecoverable_delivery_errors(error)
+      return false unless error.is_a?(Rdkafka::RdkafkaError)
+      return false unless error.code == :msg_timed_out # -192
```

Comment:
So this code _only_ handles timeouts, or does `msg_timed_out` cover any reason for error?

Reply:
The timeout is a variant of permanent failure, [in upstream's words](https://github.com/edenhill/librdkafka/blob/master/INTRODUCTION.md#error-local-time-out):
> The message could not be successfully transmitted before message.timeout.ms expired, typically due to no leader being available or no broker connection. The message may have been retried due to other errors but those error messages are abstracted by the ERR__MSG_TIMED_OUT error code.

Meaning that librdkafka does not report any transient errors on the delivery callback. As per [this comment](https://github.com/edenhill/librdkafka/issues/1066#issuecomment-280776570), we have three+one scenarios:
1. success
2. librdkafka gave up due to whatever reason.
3. broker rejects (e.g. message too large), [see upstream docs](https://github.com/edenhill/librdkafka/blob/master/INTRODUCTION.md#error-permanent-errors).
4. `Rdkafka::AbstractHandle::WaitTimeoutError` is introduced by rdkafka-ruby, and returns after 60s iirc, even if the failure is not final (i.e. scenario 2)

1 is obviously a no op, 3 is unrecoverable unless the client/caller/implementer changes their message, 2 is this PR. Not all errors that cause 2 can be solved by reconnecting, but a stuck broker connection might be fixed that way. 4 is potentially something racecar can just hide (as in `log.Info "still trying to deliver message..."`) until 2 occurs.

It should be easy enough to wrap these with an extended error message, pointing to the relevant upstream docs and kafka or racecar settings.

For `Rdkafka::AbstractHandle::WaitTimeoutError` the most sensible options seems to be to not have the error message in the first place on high level consumers by setting a wait time > `message.timeout.ms`. The other option would be to try to log something helpful like the message key, but then again we suspect it's not specifically that message but something "on infrastructure level". Plus printing the key if it's a megabyte of JSON also has its own set of problems. So the only reliable takeaway would be that the Ruby process is not stuck… which one can have in a better way from [rbspy](https://rbspy.github.io/).

Reply:
Awesome! It would also be _amazing_ if you could put this knowledge into the code as a comment, since it's not at all obvious how that works.